File: Pod/Classes/UINavigationController+TestHelpers.m
//
//  UINavigationController+TestHelpers.m
//  Pods
//
//  Created by Anthony Miller on 4/19/17.
//
//

#import <objc/runtime.h>

#import <JRSwizzle/JRSwizzle.h>

@interface UINavigationController (TestHelpers)
@property (nonatomic, strong) NSArray* AM_testViewControllers;
@property (readonly, nonatomic, nullable) UIViewController* AM_testTopViewController;
@end

@implementation UINavigationController (TestHelpers)
@dynamic AM_testViewControllers;

+ (void)load {
    static dispatch_once_t onceToken;
    dispatch_once(&onceToken, ^{
 
        NSError *error;
        [self jr_swizzleMethod:@selector(pushViewController:animated:)
                    withMethod:@selector(AM_testPushViewController:animated:)
                         error:&error];
        
        [self jr_swizzleMethod:@selector(popViewControllerAnimated:)
                    withMethod:@selector(AM_testPopViewControllerAnimated:)
                         error:&error];

        [self jr_swizzleMethod:@selector(popToViewController:animated:)
                    withMethod:@selector(AM_testPopToViewController:animated:)
                         error:&error];
        
        [self jr_swizzleMethod:@selector(popToRootViewControllerAnimated:)
                    withMethod:@selector(AM_testPopToRootViewControllerAnimated:)
                         error:&error];
        
        [self jr_swizzleMethod:@selector(viewControllers)
                    withMethod:@selector(AM_testViewControllers)
                         error:&error];
        
        [self jr_swizzleMethod:@selector(setViewControllers:)
                    withMethod:@selector(setAM_testViewControllers:)
                         error:&error];
        
        [self jr_swizzleMethod:@selector(topViewController)
                    withMethod:@selector(AM_testTopViewController)
                         error:&error];
        
        if (error != nil) {
            NSLog(@"%@", error);
        }
    });
}

- (void)AM_testPushViewController:(UIViewController *)viewController animated:(BOOL)animated {
    NSArray *viewControllers = [self.viewControllers arrayByAddingObject:viewController];
    [self addChildViewController:viewController];
    self.viewControllers = viewControllers;
}

- (nullable UIViewController *)AM_testPopViewControllerAnimated:(BOOL)animated {
    UIViewController *poppedVC = self.viewControllers.lastObject;
    [poppedVC removeFromParentViewController];
    
    NSMutableArray *viewControllers = [NSMutableArray arrayWithArray:self.viewControllers];
    [viewControllers removeLastObject];
    self.viewControllers = viewControllers;
    
    return poppedVC;
}

- (nullable NSArray<UIViewController *> *)AM_testPopToViewController:(UIViewController *)viewController animated:(BOOL)animated {
    NSMutableArray *VCs = [NSMutableArray arrayWithArray:self.viewControllers];
    
    BOOL foundVC = false;
    NSMutableArray *poppedVCs = [NSMutableArray array];
    
    while (!foundVC) {
        UIViewController *VC = VCs.lastObject;
        if (VC == viewController) {
            foundVC = true;
            
        } else {
            [VCs removeLastObject];
            [VC removeFromParentViewController];
            [poppedVCs addObject:VC];
        }
    }
    
    self.viewControllers = VCs;
    return poppedVCs;
}

- (nullable NSArray<UIViewController *> *)AM_testPopToRootViewControllerAnimated:(BOOL)animated {
    if (self.viewControllers.count <= 1) { return nil; }

    UIViewController *rootVC = self.viewControllers.firstObject;
    
    NSMutableArray *oldVCs = [NSMutableArray arrayWithArray:self.viewControllers];
    [oldVCs removeObjectAtIndex:0];
    
    self.viewControllers = @[rootVC];
    
    for (UIViewController *VC in oldVCs) {
        [VC removeFromParentViewController];
    }
    
    return [NSArray arrayWithArray:oldVCs];
}

- (id)AM_testViewControllers {
    UIViewController *rootVC = self.AM_testViewControllers.firstObject;
    NSArray *VCs = (NSArray *)objc_getAssociatedObject(self, @selector(AM_testViewControllers));
    
    if (VCs != nil) {
        if (rootVC != nil && VCs.firstObject != rootVC) {
            NSMutableArray *mutableVCs = [NSMutableArray arrayWithArray:VCs];
            [mutableVCs insertObject:rootVC atIndex:0];
            return mutableVCs;
            
        } else {
            return VCs;
        }
    }
    
    if (rootVC != nil) {
        objc_setAssociatedObject(self, @selector(AM_testViewControllers), @[rootVC], OBJC_ASSOCIATION_RETAIN_NONATOMIC);
        return @[rootVC];
        
    } else {
        objc_setAssociatedObject(self, @selector(AM_testViewControllers), @[], OBJC_ASSOCIATION_RETAIN_NONATOMIC);
        return @[];
    }
}

- (void)setAM_testViewControllers:(NSArray<UIViewController *> *)newValue {
        for (UIViewController *viewController in self.viewControllers) {
            [viewController removeFromParentViewController];
        }
        for (UIViewController *viewController in newValue) {
            [self addChildViewController:viewController];
        }
        objc_setAssociatedObject(self,
                                 @selector(AM_testViewControllers),
                                 newValue,
                                 OBJC_ASSOCIATION_RETAIN_NONATOMIC);
}

- (nullable UIViewController*)AM_testTopViewController {
    return self.viewControllers.lastObject;
}

@end

File: Pod/Classes/UINavigationController+TestHelpers.swift
//
//  UINavigationController+TestHelpers.swift
//  AMTestHelpers
//
//  Created by Anthony Miller on 10/30/15.
//

import UIKit

import JRSwizzle

/*
 * MARK: Swizzle UINavigationController
 */

fileprivate let swizzleUINavigationControllerPushViewController: () = {
    let originalSelector = #selector(UINavigationController.pushViewController(_:animated:))
    let swizzledSelector = #selector(UINavigationController.AM_testPushViewController(_:animated:))
    
    do { try UINavigationController.self.jr_swizzleMethod(originalSelector, withMethod: swizzledSelector) } catch {
        debugPrint(error)
    }
}()

fileprivate let swizzleUINavigationControllerPopViewController: () = {
    let originalSelector = #selector(UINavigationController.popViewController(animated:))
    let swizzledSelector = #selector(UINavigationController.AM_testPopViewControllerAnimated(_:))
    
    do { try UINavigationController.self.jr_swizzleMethod(originalSelector, withMethod: swizzledSelector) } catch {
        debugPrint(error)
    }
}()

fileprivate let swizzleUINavigationControllerPopToViewController: () = {
    let originalSelector = #selector(UINavigationController.popToViewController(_:animated:))
    let swizzledSelector = #selector(UINavigationController.AM_testPopToViewController(_:animated:))
    
    do { try UINavigationController.self.jr_swizzleMethod(originalSelector, withMethod: swizzledSelector) } catch {
        debugPrint(error)
    }
}()

fileprivate let swizzleUINavigationControllerPopToRootViewController: () = {
    let originalSelector = #selector(UINavigationController.popToRootViewController(animated:))
    let swizzledSelector = #selector(UINavigationController.AM_testPopToRootViewControllerAnimated(_:))
    
    do { try UINavigationController.self.jr_swizzleMethod(originalSelector, withMethod: swizzledSelector) } catch {
        debugPrint(error)
    }
}()

fileprivate let swizzleUINavigationControllerViewControllersProperty: () = {
    let originalGetter = #selector(getter: UINavigationController.viewControllers)
    let swizzledGetter = #selector(getter: UINavigationController.AM_testViewControllers)
    
    do { try UINavigationController.self.jr_swizzleMethod(originalGetter, withMethod: swizzledGetter) } catch {
        debugPrint(error)
    }
    
    let originalSetter = #selector(setter: UINavigationController.viewControllers)
    let swizzledSetter = #selector(setter: UINavigationController.AM_testViewControllers)
    
    do { try UINavigationController.self.jr_swizzleMethod(originalSetter, withMethod: swizzledSetter) } catch {
        debugPrint(error)
    }
    
}()

fileprivate let swizzleUINavigationControllerTopViewController: () = {
    let originalSelector = #selector(getter: UINavigationController.topViewController)
    let swizzledSelector = #selector(getter: UINavigationController.AM_testTopViewController)
    
    do { try UINavigationController.self.jr_swizzleMethod(originalSelector, withMethod: swizzledSelector) } catch {
        debugPrint(error)
    }
}()

extension UINavigationController {
    open override class func initialize() {
        swizzleUINavigationControllerPushViewController
        swizzleUINavigationControllerPopViewController
        swizzleUINavigationControllerPopToViewController
        swizzleUINavigationControllerPopToRootViewController
        swizzleUINavigationControllerViewControllersProperty
        swizzleUINavigationControllerTopViewController
    }
    
    func AM_testPushViewController(_ viewController: UIViewController, animated: Bool) {
        var viewControllers = self.viewControllers
        viewControllers.append(viewController)
        self.addChildViewController(viewController)
        self.viewControllers = viewControllers
    }
    
    func AM_testPopViewControllerAnimated(_ animated: Bool) -> UIViewController? {
        let poppedVC = self.viewControllers.last
        poppedVC?.removeFromParentViewController()
        
        self.viewControllers = Array(self.viewControllers.dropLast())
        return poppedVC
    }
    
    func AM_testPopToViewController(_ viewController: UIViewController, animated: Bool) -> [UIViewController]? {
        var VCs = self.viewControllers
        var foundVC = false
        var poppedVCs: [UIViewController] = []
        
        while !foundVC {
            let VC = VCs.last
            if VC === viewController {
                foundVC = true
                
            } else {
                let poppedVC = VCs.removeLast()
                poppedVC.removeFromParentViewController()
                poppedVCs.append(poppedVC)
            }
        }
        
        self.viewControllers = VCs
        return poppedVCs
    }
    
    func AM_testPopToRootViewControllerAnimated(_ animated: Bool) -> [UIViewController]? {
        guard let rootVC = self.viewControllers.first,
            self.viewControllers.count > 1 else { return nil }
        
        let oldVCs = self.viewControllers.dropFirst()
        self.viewControllers = [rootVC]
        
        for VC in oldVCs {
            VC.removeFromParentViewController()
        }
        
        return Array(oldVCs)
    }
    
    fileprivate struct AssociatedKeys {
        static var TestViewControllers = "AM_testViewControllers"
    }
    
    var AM_testViewControllers: [UIViewController] {
        get {
            guard let VCs = objc_getAssociatedObject(self,
                                                     &AssociatedKeys.TestViewControllers) as? [UIViewController] else {
                                                        
                                                        if let rootVC = AM_testViewControllers.first {
                                                            objc_setAssociatedObject(self,
                                                                                     &AssociatedKeys.TestViewControllers,
                                                                                     [rootVC],
                                                                                     objc_AssociationPolicy.OBJC_ASSOCIATION_RETAIN_NONATOMIC)
                                                            
                                                            return [rootVC]
                                                        } else {
                                                            objc_setAssociatedObject(self,
                                                                                     &AssociatedKeys.TestViewControllers,
                                                                                     [],
                                                                                     objc_AssociationPolicy.OBJC_ASSOCIATION_RETAIN_NONATOMIC)
                                                            return []
                                                        }
            }
            
            if let rootVC = AM_testViewControllers.first, VCs.first != rootVC {
                return [rootVC] + VCs
            }
            
            return VCs
        }
        set {
            for viewControllers in self.viewControllers {
                viewControllers.removeFromParentViewController()
            }
            for viewController in newValue {
                self.addChildViewController(viewController)
            }
            objc_setAssociatedObject(self,
                                     &AssociatedKeys.TestViewControllers,
                                     newValue,
                                     objc_AssociationPolicy.OBJC_ASSOCIATION_RETAIN_NONATOMIC)
        }
    }
    
    var AM_testTopViewController: UIViewController? {
        get {
            return self.viewControllers.last
        }
    }
    
}
